File: Python/Path Sum II.py
# Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.

# For example:
# Given the below binary tree and sum = 22,
              # 5
             # / \
            # 4   8
           # /   / \
          # 11  13  4
         # /  \    / \
        # 7    2  5   1
# return
# [
   # [5,4,11,2],
   # [5,8,4,5]
# ]


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
	def pathSum(self, root, Sum):
		currentSum = 0
		current = []
		solution = []
		if root != None:
			self.pathSolution(root,Sum,current,solution,currentSum)
		return solution
	def pathSolution(self,root,Sum,current,solution,currentSum):
		currentSum = currentSum + root.val
		current.append(root.val)
		if currentSum == Sum and root.left == None and root.right == None:
			solution.append(list(current))
		else:
			if root.left != None:
				self.pathSolution(root.left,Sum,current,solution,currentSum)
			if root.right != None:
				self.pathSolution(root.right,Sum,current,solution,currentSum)
		current.pop()
		currentSum = currentSum - root.val
		return

File: Java/Path_Sum_II.java
class Path_Sum_II{
	public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<ArrayList<Integer>>();
        List<Integer> solution = new ArrayList<Integer>();
        if(root == null)
        	return res;
        genPath(root,0,sum,res,solution);
        //genPath(root,root.val,sum,res,solution);
        return res;
    }
    private void genPath(TreeNode root, int curr, int sum, List<List<Integer>> res,List<Integer> solution){
    	curr += root.val;
    	solution.add(root.val);
    	if(root.left == null && root.right == null && curr == sum){
    		List<Integer> temp = new ArrayList<Integer>(solution);
    		res.add(temp);
    		solution.remove(solution.size()-1);
    		return;
    	}
    	if(root.left != null){
    		genPath(root.left,curr,sum,res,solution);
    	}
    	if(root.right != null){
    		genPath(root.right,curr,sum,res,solution);
    	}
    	solution.remove(solution.size()-1);
    }private void genPath(TreeNode root, int curr, int sum, List<List<Integer>> res,List<Integer> solution){
    	if(root.left == null && root.right == null && curr == sum){
    		List<Integer> temp = new ArrayList<Integer>(solution);
    		res.add(temp);
    		return;
    	}
    	if(root.left != null){
    		solution.add(root.left);
    		genPath(root.left,curr+root.left.val,sum,res,solution);
    		solution.remove(solution.size()-1);
    	}
    	if(root.right != null){
    		solution.add(root.left);
    		genPath(root.right,curr+root.right.val,sum,res,solution);
    		solution.remove(solution.size()-1);
    	}
    }
}